### PR TITLE
Temporarily mask bad modules on DLS I03 Eiger

### PR DESCRIPTION
**Pull request by rjgildea**
Time-limited masking from 2021_04_22 till 2021_05_03 (inclusive).

### CODE REVIEW

File: format/FormatNexusEigerDLS16MI03.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    return get_pilatus_timestamp(timestamp)
+
+
+class FormatNexusEigerDLS16MI03(FormatNexusEigerDLS16M):
+    @staticmethod
+    def understand(image_file):
+        # Get the file handle
+        with h5py.File(image_file, "r") as handle:
+            name = h5str(FormatNexusEigerDLS16M.get_instrument_name(handle))
+            return name and name.upper() in {"DIAMOND BEAMLINE I03", "DLS I03"}
```
Comment (graeme-winter) [inline format/FormatNexusEigerDLS16MI03.py:22]: Please could we add the datestamp gating in the `understand` method -> we don't need to even consider using this format method for pre-issue or post-issue data

File: format/FormatNexusEigerDLS16MI03.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        ny = 512  # module pixels y
+        dx = 12  # module gap size
+        dy = 38  # module gap size
+        timestamp = get_start_time(self._image_file)
+        if timestamp > calendar.timegm(
+            (2021, 4, 22, 0, 0, 0)
+        ) and timestamp < calendar.timegm((2021, 5, 4, 0, 0, 0)):
+            # 2021 run 2
+            # Full module
+            bad_modules = {(2, 5), (3, 5)}
```
Comment (graeme-winter) [inline format/FormatNexusEigerDLS16MI03.py:42]: Please could you add a comment block above explaining which modules are which (I know, but others won't)

File: format/FormatNexusEigerDLS16MI03.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+from dxtbx.format.nexus import h5str
+
+
+def get_start_time(filename):
+    with h5py.File(filename, mode="r") as f:
+        timestamp = h5str(f["/entry/start_time"][()]).rstrip("Z")
+    return get_pilatus_timestamp(timestamp)
+
+
+class FormatNexusEigerDLS16MI03(FormatNexusEigerDLS16M):
```
Comment (graeme-winter) [inline format/FormatNexusEigerDLS16MI03.py:16]: Please can you add a docstring explaining the context of this e.g. 

"Specific format overloading the module behaviour for the Eiger 2XE 16M on i03 for dates..."

For future reference

File: format/FormatNexusEigerDLS16MI03.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        dy = 38  # module gap size
+        timestamp = get_start_time(self._image_file)
+        if timestamp > calendar.timegm(
+            (2021, 4, 22, 0, 0, 0)
+        ) and timestamp < calendar.timegm((2021, 5, 4, 0, 0, 0)):
+            # 2021 run 2
+            # Full module
+            bad_modules = {(2, 5), (3, 5)}
+            for i, j in bad_modules:
+                self._detector_model[0].add_mask(
```
Comment (graeme-winter) [inline format/FormatNexusEigerDLS16MI03.py:44]: Have you also considered impact on ROI data e.g. central 8 modules (which one of these modules intersects with) 

I suspect you need a check which verifies the mode and then applies the correct patch

File: format/FormatNexusEigerDLS16MI03.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        self._image_file = image_file
+
+    def _start(self):
+        super()._start()
+        nx = 1028  # module pixels x
+        ny = 512  # module pixels y
+        dx = 12  # module gap size
+        dy = 38  # module gap size
+        timestamp = get_start_time(self._image_file)
+        if timestamp > calendar.timegm(
```
Comment (graeme-winter) [inline format/FormatNexusEigerDLS16MI03.py:37]: Please could you move this to `understand()`

**Review by graeme-winter — CHANGES_REQUESTED** [review]
Thank you for the speedy resolution - couple of suggestions which I think need addressing before we merge this.

File: format/FormatNexusEigerDLS16MI03.py (hunk 2 above)
Comment (rjgildea) [reply format/FormatNexusEigerDLS16MI03.py:39]: I've modified the variable names to `i_row` and `i_col` to hopefully help make the code clearer.

Comment (graeme-winter) [reply format/FormatNexusEigerDLS16MI03.py:39]: Good enough :)

File: format/FormatNexusEigerDLS16MI03.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    def _start(self):
+        super()._start()
+        nx = 1028  # module pixels x
+        ny = 512  # module pixels y
+        dx = 12  # module gap size
+        dy = 38  # module gap size
+        if self._detector_model[0].get_image_size() == (4148, 4362):
+            # Full detector mode
+            bad_modules = {(2, 5), (3, 5)}
```
Comment (graeme-winter) [inline format/FormatNexusEigerDLS16MI03.py:39]: Verified correct

File: format/FormatNexusEigerDLS16MI03.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            for i_col, i_row in bad_modules:
+                self._detector_model[0].add_mask(
+                    (nx + dx) * i_col,
+                    (ny + dy) * i_row,
+                    (nx + dx) * i_col + nx,
+                    (ny + dy) * i_row + ny,
+                )
+        else:
+            # ROI detector mode
+            i_col = 1
```
Comment (graeme-winter) [inline format/FormatNexusEigerDLS16MI03.py:49]: Verified correct, though I would wonder whether it is helpful to follow the same structure as above if we happen to revisit this file? I.e. a list of one module to patch

File: format/FormatNexusEigerDLS16MI03.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    ROI mode.
+    """
+
+    @staticmethod
+    def understand(image_file):
+        # Get the file handle
+        with h5py.File(image_file, "r") as fh:
+            name = h5str(FormatNexusEigerDLS16M.get_instrument_name(fh))
+            if name and name.upper() in {"DIAMOND BEAMLINE I03", "DLS I03"}:
+                timestamp = get_pilatus_timestamp(
```
Comment (graeme-winter) [inline format/FormatNexusEigerDLS16MI03.py:24]: 👍  thank you

**Review by graeme-winter — APPROVED** [review]
Looks good, thanks for the revisions (in my eyes, improvements)

File: format/FormatNexusEigerDLS16MI03.py (hunk 7 above)
Comment (rjgildea) [reply format/FormatNexusEigerDLS16MI03.py]: I've simplified the if/else statement here: https://github.com/cctbx/dxtbx/pull/348/commits/bff85c365f958046b958b4e21728bedccb88d189

Comment (graeme-winter) [reply format/FormatNexusEigerDLS16MI03.py]: Noted, saw the commit, 👍

**Pull request merged**